filecoin-proofs/src/bin/paramfetch.rs

```diff
@@ -113,18 +115,38 @@ Set http_proxy/https_proxy environment variables to specify proxy for ipfs gatew
 }
 
 fn fetch(matches: &ArgMatches) -> Result<()> {
-    let json_path = PathBuf::from(matches.value_of("json").unwrap_or("./parameters.json"));
+    let manifest = if matches.is_present("json") {
+        let json_path = PathBuf::from(matches.value_of("json").unwrap());
+        println!("using json file: {:?}", json_path);
+
+        if !json_path.exists() {
+            return Err(err_msg(format!(
+                "json file '{}' does not exist",
+                &json_path.to_str().unwrap_or("")
+            )));
+        }
+
+        let file = File::open(&json_path)?;
+        let reader = BufReader::new(file);
+        let result = serde_json::from_reader(reader);
+
+        if result.is_err() {
+            return Err(err_msg(format!(
+                "json file '{}' did not parse correctly: {}",
+                &json_path.to_str().unwrap_or(""),
+                result.err().unwrap(),
+            )));
+        }
+
+        result.unwrap()
+    } else {
+        println!("using built-in manifest");
+        serde_json::from_str(&DEFAULT_PARAMETERS)?
+    };
```

Comment:
See note above. If you use `map_err`, then you can use `?` after the `else` block, e.g.

```rust
let manifest = if matches.is_present("json") {
    serde_json::from_reader(reader).map_err(|err| {
        Err(err_msg(format!( // etc
    })
} else {
    println!("using built-in manifest");
    serde_json::from_str(&DEFAULT_PARAMETERS)
}?;

```